syntax = "proto3";

package musicqueue;

service QueueService {
  rpc AddTrack (Track) returns (QueueResponse);
  rpc RemoveTrack (TrackAction) returns (QueueResponse);
  rpc VoteTrack (VoteRequest) returns (QueueResponse);
  rpc GetQueue (Empty) returns (QueueList);
  rpc GetMetadata (TrackAction) returns (Track);
  rpc PlayNext (Empty) returns (Track);
  rpc GetHistory (Empty) returns (QueueList);
}

message Track {
  int32 id = 1;
  string title = 2;
  string artist = 3;
  int32 duration = 4;
  int32 votes = 5;
}

message TrackAction {
  int32 id = 1;
}

message VoteRequest {
  int32 id = 1;
  bool up = 2;
}

message QueueResponse {
  string message = 1;
  repeated Track queue = 2;
}

message QueueList {
  repeated Track queue = 1;
}

message Empty {}
